### ISSUE TITLE
Sparkle creates broken app bundles for some users

### ISSUE THREAD

**Issue by fzwo** [sparkle-project/Sparkle#441]
We have the strangest problem.

We have a relatively large app, and ship mono within the app bundle. The .app is zipped for Sparkle (~250 MB zip, 600+ MB app), and is signed.

We've had reports from several (but not all) users that the app is broken after an update. We were scratching our heads because all the nib files seemed to be missing. SparkleUpdateLogs were always clean, and we couldn't find anything in our own crash logs (both through Hockey and the log files on the machine) except for the mysteriously missing xib files.

We were now able to take a closer look at one of the broken apps. Sure enough, the nibs were missing, as was everything else starting with S (the first letter of our company prefix) in the bundle's resources folder. Actually, everything after M was missing. Or rather, everything after Mono. Digging deeper, it seems that sometime during copying mono.framework, the updater stopped.

This resulted in an app bundle that reports as broken when trying to run on a Mac with Gatekeeper on, because the signature does not match anymore. On a Mac with gatekeeper off, it's more insidious: The app starts, and runs, but certain Windows won't open because their nibs can't be found.

This has happened to a handful of users now, some of them multiple times. We'd really love to find out what's causing this, and how to protect against it. Any ideas? (We're on commit https://github.com/sparkle-project/Sparkle/commit/7e1b47976dd80c09abd9f772a4117eeb08dc15f1 - there haven't been any functional changes since then)

**Comment by steiger** [comment]
FWIW, another log file excerpt and crash report (https://gist.github.com/steiger/380f99126cd52f5c53e8).

14.02.15 08:48:19.258	Mailplane 3	Sparkle: ===== Mailplane 3 =====
14.02.15 08:50:59.334	Mailplane 3	Sparkle: Extracting /Users/jo/Library/Application Support/Mailplane 3/.Sparkle/Mailplane 3 1319/Mailplane_3_1319.tbz using 'tar -jxC "$DESTINATION"'
14.02.15 08:51:38.805	ReportCrash	Saved crash report for Autoupdate[5109] version 1.8.0 (1.8.0) to /Users/jo/Library/Logs/DiagnosticReports/Autoupdate_2015-02-14-085138_betty.crash
14.02.15 09:59:20.719	Mailplane 3	Unable to load nib file: MainMenu, exiting

**Comment by pornel** [comment]
I've recently merged 65238c35a9dec983f8c49f3475c2cfd5f926f3f2. Does that fix the nib loading problems for you?

**Comment by fzwo** [comment]
Pornel, I don't think that is the same problem that we are experiencing. To be clear, the updated application bundle is missing some files after update, including the nibs. It's not a nib loading problem - the nibs are not there. We test this ourselves in separate code when launching the app.

**Comment by iwankerl** [comment]
Perhaps the Autoupdate app should check the code signature of the app towards the end of +[SUPlainInstaller copyPathWithAuthentication:overPath:temporaryName:error:], just before we release from quarantine?

**Comment by fzwo** [comment]
That sounds like a good idea. In any case, if it passes here and the problem still happens, that would help find out where/why it's happening. (And, by the way, it still does - our latest Sparkle at customers is forked from 729087f696a9a1aee6d660b5df39f4faa888ffd1)

**Comment by zorgiepoo** [comment]
I don't think Sparkle right now checks for nested items/resources in the app bundle when it's checking if the code signature is correct (see kSecCSCheckNestedCode). So Sparkle's validation may not work for the entire bundle.

**Comment by zorgiepoo** [comment]
Moreover, from what I can tell even if the code signature is broken, if you provide a valid DSA signature for the archive, it'll still continue on. Although, doesn't appear to be the case here I don't think.

**Comment by jhammer** [comment]
We are seeing this same issue on several customer installs. It happens across 10.7 to 10.10. The files that go missing from the app bundle vary from customer to customer, too.

**Comment by zorgiepoo** [comment]
The crash looks like a memory management issue from the crash logs I've seen submitted here. The problem doesn't look like it's with copying because the thread that does the copying doesn't cause the crash, although the crash happens to occur around the same time. I am also assuming when this issue occurs, that Sparkle does not automatically restart the application? I pushed the commit that @greenbuffo suggested doing early, though this is only a guess of it being the issue.

**Comment by pornel** [comment]
I've rewritten unarchiving code 06c6a3b7f2e55ad10f87c28a79f88a49741c89ac

It now uses higher-level code to pipe streams, and uses absolute paths to the tools (which should cover cases where somebody installed broken tar or ditto).

**Comment by fzwo** [comment]
Thank you, @pornel !
I've merged the changes into our branch, which will be shipped with next week's release. It will likely be two weeks at least before we release another version and can see if this resolves the issue.

**Comment by bdkjones** [comment]
What's the status of this issue?

I've had about 60-70 reports of this corruption. My app is using Sparkle 1.9, which appears to be the latest official release. Is this fix incorporated into the 1.10 pre-release? If so, can you guys push an official release so that we can fix our apps? 

This is a pretty big problem to just leave outstanding for months.

**Comment by bdkjones** [comment]
Anything?

**Comment by pornel** [comment]
As noted in https://github.com/sparkle-project/Sparkle/issues/441#issuecomment-95376311 it may have been fixed. 

Do you see the problem in 1.10rc1 still?

**Comment by pornel** [comment]
@fzwo have you got any feedback from 1.10rc1 yet?

**Comment by bdkjones** [comment]
Well, the only way to test that would be to ship an update to my app using a pre-release version of Sparkle. That makes me really uncomfortable because Sparkle is *THE* thing I rely on. If Sparkle breaks, then thousands and thousands of users will be permanently stranded on an old version of my app because I'll have no way to notify them of future updates. The updater is not something I want to beta test in a production app.

**Comment by zorgiepoo** [comment]
While it could be an issue with unarchiving the application, I am sort of in doubt about that. The crash reports I've seen here seem to be related with Sparkle crashing on the main thread, while it's copying the app bundle to its destination on a different thread. @greenbuffo's commit looks like the closest thing to me that might have fixed the issue, though @fzwo states it is still an issue after that commit (but hasn't given logs or crash reports since then). @fzwo's build with the log insertions also implies that the copying operation never finished. I also presume @greenbuffo's commit never got into an official release yet.

If you have crash reports from Sparke's helper app Autoupdate.app (since it's a helper tool it may not show up in things like HockeyApp!) and/or logs from the Console, that is probably helpful.

**Comment by bdkjones** [comment]
Unfortunately, I have not been able to reproduce the issue on my machine, so I don't have any crash logs to send. From my experience, the majority of updates succeed successfully. I've had about 50-60 reports of the corruption out of many thousands of updates. It's definitely enough to be a problem.

One user suggested that the problem occurs if the application being updated is not in `/Applications`, but I can't confirm that. It may be worth testing, though.

**Comment by fzwo** [comment]
Inconclusive. We've had customers who were updating last week and were still experiencing the issue, but it's not clear which version they updated from. I'll see if I can dig up some tickets where that's clear.

If they updated from the penultimate version to the most recent one, they were using our fork from 63052aa3d11759c8e960d4fb2757c5336e9ccaf9 (which I think contains most of the fixes).

At least one of those users also had their app residing in /Applications.

**Comment by jhammer** [comment]
I'd be interested to know if @fzwo has seen this problem go away in their fork (which I think has the fixes from @pornel and @greenbuffo that are mentioned above). Our fork, which includes only the fixes @greenbuffo, is still crashing and burning. 10.7 to 10.11. Any other feedback or reports would be welcome, too.

**Comment by bdkjones** [comment]
I've had about 25,000 updates from a version of my app with Sparkle `1.10.0.rc1` TO a version of my app with Sparkle `1.10.0.rc1`. I've not had any reports of crashes.

The issue appears to be isolated to the 1.9 version.

I know we're all human, but Sparkle is sort of like video drivers: it just HAS to work. If Sparkle breaks, users get stranded on old versions of apps and developers can't do anything about it. The fact that this issue made it into a release build should be a call for stronger testing. You guys are writing mission-critical software!

**Comment by pornel** [comment]
@bdkjones Thank you for checking.

I agree that Sparkle is mission-critical. How do you recommend we test it?

**Comment by jakepetroules** [comment]
@pornel #578 would be a good step in the right direction - have you had a chance to take a look yet?

**Comment by zorgiepoo** [comment]
#575 would also the basis for making corrupt updates impossible (and much better error checking) = ).

**Comment by zorgiepoo** [comment]
Regardless you *need* to use Sparkle's DSA signatures if you want to ensure your download is correct (in rare cases http corrupts on you for example; also a security issue). I agree with Jake about looking at the test app patch first. I could possibly look into adding other options later to it, like a "big update" or "delta update" ones for example.

**Comment by bdkjones** [comment]
Actually: scratch that. 

I woke up this morning to about 15 emails from people that updated and found the app corrupted and unable to launch. Sparkle is still broken, it seems. These were 1.10.0.rc1 to 1.10.0.rc1 updates. The crash reports showed missing assets like fonts, images, etc.

I'm going to roll my app back to the last GOOD version of Sparkle. I wish I had more information to help you guys diagnose and debug this. Unfortunately, I don't think Sparkle is safe to use past the last ancient release until there's solid testing in place.

**Comment by tonglil** [comment]
```
Dyld Error Message:
  Library not loaded: @loader_path/../Frameworks/Sparkle.framework/Versions/A/Sparkle
  Referenced from: /Applications/- Tools/Something.app/Contents/MacOS/Something
  Reason: no suitable image found.  Did find:
	/Applications/- Tools/Something.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle: code signature invalid for '/Applications/- Tools/Something.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle'

	/Applications/- Tools/Something.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle: code signature invalid for '/Applications/- Tools/Something.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle'
```

Sparkle is common theme in all of the apps on my computer that will not open.

Any workarounds?

**Comment by jakepetroules** [comment]
As the error message says, the code signature for the Sparkle framework is invalid. When you embed the Sparkle framework in your application in Xcode, you should ensure that "code sign on copy" is checked.

You can also run `codesign --deep -vvvv '/Applications/- Tools/Something.app'` to help debug code signing problems.

**Comment by tonglil** [comment]
@jakepetroules mm, I tried the same thing, it says valid in terminal, but invalid when I try to open the app.

This all started happening after the update of OS X 10.10.5.

```
codesign --deep -vvvv BetterTouchTool.app
--prepared:/Applications/- Tools/BetterTouchTool.app/Contents/Frameworks/ShortcutRecorder.framework/Versions/Current/.
--prepared:/Applications/- Tools/BetterTouchTool.app/Contents/Frameworks/FeedbackReporter.framework/Versions/Current/.
--validated:/Applications/- Tools/BetterTouchTool.app/Contents/Frameworks/ShortcutRecorder.framework/Versions/Current/.
--validated:/Applications/- Tools/BetterTouchTool.app/Contents/Frameworks/FeedbackReporter.framework/Versions/Current/.
--prepared:/Applications/- Tools/BetterTouchTool.app/Contents/Frameworks/Sparkle.framework/Versions/Current/.
--validated:/Applications/- Tools/BetterTouchTool.app/Contents/Frameworks/Sparkle.framework/Versions/Current/.
--prepared:/Applications/- Tools/BetterTouchTool.app/Contents/MacOS/libLeap.dylib
--validated:/Applications/- Tools/BetterTouchTool.app/Contents/MacOS/libLeap.dylib
BetterTouchTool.app: valid on disk
BetterTouchTool.app: satisfies its Designated Requirement
```

Open app:
```
Process:               BetterTouchTool [2517]
Path:                  /Applications/- Tools/BetterTouchTool.app/Contents/MacOS/BetterTouchTool
Identifier:            com.hegenberg.BetterTouchTool
Version:               ???
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           BetterTouchTool [2517]
User ID:               501

Date/Time:             2015-08-31 14:42:42.504 -0700
OS Version:            Mac OS X 10.10.5 (14F27)
Report Version:        11
Anonymous UUID:        A882AD3B-D354-6926-5F94-B210A7A2D68C

Sleep/Wake UUID:       57C9E931-1A6F-42CD-A63A-E3908E01CAD8

Time Awake Since Boot: 1500 seconds
Time Since Wake:       340 seconds

Crashed Thread:        0

Exception Type:        EXC_BREAKPOINT (SIGTRAP)
Exception Codes:       0x0000000000000002, 0x0000000000000000

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  Library not loaded: @rpath/Sparkle.framework/Versions/A/Sparkle
  Referenced from: /Applications/- Tools/BetterTouchTool.app/Contents/MacOS/BetterTouchTool
  Reason: no suitable image found.  Did find:
	/Applications/- Tools/BetterTouchTool.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle: code signature invalid for '/Applications/- Tools/BetterTouchTool.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle'

Binary Images:
    0x7fff68953000 -     0x7fff68989887  dyld (353.2.3) <B1B370A5-479F-3533-8AD7-97B687D4F989> /usr/lib/dyld
    0x7fff95339000 -     0x7fff95385fff  com.apple.corelocation (1486.17 - 1615.24) <8825B3E2-E053-3E01-AE31-793443962D06> /System/Library/Frameworks/CoreLocation.framework/Versions/A/CoreLocation
    0x7fff96c92000 -     0x7fff96d76fff  libcrypto.0.9.8.dylib (52.40.1) <D123FC57-E2EB-3E08-A51B-6C317EA83EB5> /usr/lib/libcrypto.0.9.8.dylib
    0x7fff972b0000 -     0x7fff97352fff  com.apple.Bluetooth (4.3.6 - 4.3.6f3) <E5DC1699-2727-329C-A843-C80CE9A6CD7B> /System/Library/Frameworks/IOBluetooth.framework/Versions/A/IOBluetooth
```

Run codesign on specific framework:
```
codesign --deep -vvvv '/Applications/- Tools/BetterTouchTool.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle'
/Applications/- Tools/BetterTouchTool.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle: valid on disk
/Applications/- Tools/BetterTouchTool.app/Contents/MacOS/../Frameworks/Sparkle.framework/Versions/A/Sparkle: satisfies its Designated Requirement
```

**Comment by tonglil** [comment]
Also, computer security setting is allow apps from anywhere.

**Comment by zorgiepoo** [comment]
Maybe I'd also try `'spctl -a -t exec -vv /Applications/- Tools/BetterTouchTool.app'`

**Comment by tonglil** [comment]
@zorgiepoo of course, not sure what the outcome means (security disabled)?

```
spctl -a -t exec -vv '/Applications/- Tools/BetterTouchTool.app'
/Applications/- Tools/BetterTouchTool.app: accepted
source=Developer ID
override=security disabled
origin=Developer ID Application: Andreas Hegenberg
```

**Comment by zorgiepoo** [comment]
@tonglil  It looks fine to me. Your problem might be related to [this thread](https://discussions.agilebits.com/discussion/46560/1password-not-working-with-yosemite-10-10-5).

**Comment by fzwo** [comment]
FWIW, I spoke with our head supporter and he said that update-related issues have become very rare. I think the rare cases it still does happen in may be people updating from older versions that had bugged Sparkle updaters.

Sorry for taking so long to get back to you with an answer, but since I don't handle support myself, it's not trivial to get this info.

It looks like this issue may be fixed. If it ever happens again, I'll let you know!

**Comment by fzwo** [comment]
(Reading the comments by @bdkjones, maybe this isn't fixed yet, and I closed prematurely. I'll leave this open for the Sparkle team to decide.

**Comment by fzwo** [comment]
@tonglil I believe "security disabled" means that you have Gatekeeper disabled, as you stated you have. Try enabling it and running spctl again.

This may be an entirely unrelated issue.

Please take a look at this article from Craig Hockenberry: http://furbo.org/2015/07/23/code-signing-in-el-capitan/

Unfortunately, codesign will not always alert you to unsignable embedded stuff. Go through that stuff one-by-one as the article suggests.

We've noticed the issues he's talking about not just with 10.11, but also with more recent versions of 10.10. Seems that Gatekeeper has gotten stricter without Apple telling anybody. We fixed it by going through all our embedded frameworks and fixing the structure of our app bundle.

**Comment by zorgiepoo** [comment]
Yes I think @tonglil 's issue is unrelated (maybe should be created in a separate issue). Apple does indeed tend to improve code signing security every once in a while.

This issue should be open for now. There's a few possible things that can go wrong that I can think of:

* AutoUpdate app crashes. I've been playing with the Test Application recently and ran through it a bunch of times, and was never able to reproduce it crashing myself however. I've also refactored parts of the app in #596 to help prevent this. But I haven't been able to confidently say this has been resolved yet.

* Your download is corrupted and bypasses through Sparkle's security. Very rare but possible. Fixed in recent versions by improving code signing + DSA validation, and in 1.12 by deprecating serving over http *without* DSA. Sparkle cannot employ deep signing checks, and code signing is really not sufficient for data integrity regardless.

* Any external cause (power loss, updater crashes, user terminates updater, etc). Sparkle's installation process cannot guarantee non-corrupt installations currently. Corrupt installs should never happen even if  the updater crashes. Pending fix in #575

**Comment by bdkjones** [comment]
I'm pretty confident the problem is that third bullet point. Sparkle *needs* to guarantee non-corrupt installations. The crash reports I'm seeing from folks with corrupted installs of my app indicate that certain resources like a font or an image are missing. That's what's causing the crashes.

**Comment by tonglil** [comment]
@fzwo, @bdkjones, @zorgiepoo: the thing is, it's not just bettertouchtools that is not launching, applications including "bartender", dropbox, and shazam (unrelated issue with `use of the com.apple.developer.aps-environment entitlement is not allowed`).

**Comment by fzwo** [comment]
That makes it sound even more like an unrelated issue, unless I'm missing something.

**Comment by bdkjones** [comment]
No, this is definitely Sparkle-related. I just pushed another update and I still have people that are reporting that my app crashes after the update completes. The crash reports are all similar: image assets are missing. 

Can someone point out the last build of Sparkle that was available before it went public on GitHub here? The old version worked reliably and correctly and I'd like to roll back to it until the team gets a chance to sort out this problem.

**Comment by bdkjones** [comment]
@tonglil I don't think that's related. Apps that fail to launch after updating OS X are a different story than apps that fail to launch after Sparkle completes its update routine.

**Comment by zorgiepoo** [comment]
The next version of Sparkle (not ready yet) should be more reliable for preventing corrupted updates. If you use a current version, I recommend disabling Sparkle's Automatic download & installing of updates as I believe that was a likely cause for corrupted updates; see #629. I personally cannot really recommend a really old version with whatever other critical now-resolved issues they may have.

**Comment by pornel** [comment]
@bdkjones The releases page also has very old versions, including the 1.5b6 which was the latest for years. However, the 1.5b as-is won't run on El Capitan due to broken symlinks it had.

With improvements to unarchivers, file handling, atomic updates and system shutdown handling we're hopefully increasing reliability and getting close to fixing the issues you're seeing. Thank you for giving us feedback.

**Comment by bdkjones** [comment]
Not to sound ungrateful for everyone's work, but do you have any idea when that next release will be ready? This is a really troubling issue for me and it's making me and my app look really bad in front of my customers. 

Can you guys put together a release that includes the fixes for this rather than waiting on whatever new features still need to be implemented?

**Comment by pornel** [comment]
The fixes I've mentioned have already been merged in and we keep the master branch stable, so you can try it out.

**Comment by pornel** [comment]
We've had many fixes since that issue. Feel free to reopen if it's still a problem in the latest version.

**Comment by yingfangdu** [comment]
Hi, We keep seeing this issue. Every time when we release a new build to customers, we will get a crash that  the nib file is missing in the app bundle. 

It feels like User still can open the application during Sparkle is coping the new build.

**Comment by zorgiepoo** [comment]
Hi @yingfangdu. What version of Sparkle was being used during the update? Any other relevant info you may have would be helpful as well (size of application, % of users running into the issue..)

**Comment by yingfangdu** [comment]
The Sparkle version is 1.14.0, our App size is 128MB, and from the latest version we just deployed 0.5% users hit this issue.

**Comment by zorgiepoo** [comment]
Interesting.. After 1.12, we changed the code such that if copying is needed to be done at all (most likely when a different volume mount is involved), we would copy the new app to a temporary location first, and then move it into the final destination later. This change was made so that updates could not be installed in an incomplete manner, but perhaps somewhere else is causing the nib to be missing..

Is it always the same nib file that is missing? Do you think the app crashes right on launch because of this nib?

**Comment by yingfangdu** [comment]
The application crashes right when access the missing nib file. And through our several deployment, the missing nib file is different for each crash and the crash report says there is not such nib file in the app bundle.  

And we trace down the same user's telemetry data on the crash date, we see user reboots the application and there is no crash anymore. And this crash only reproduces around the date we deploy a new build. 

I have a question here, when Sparkle copies the application from temp folder to the destination folder, it is allowed to let user launch the application?  If yes, this might be an issue if the copy takes too long.

**Comment by zorgiepoo** [comment]
> I have a question here, when Sparkle copies the application from temp folder to the destination folder, it is allowed to let user launch the application? If yes, this might be an issue if the copy takes too long.

Sparkle moves, not copies, the application from the temp folder to the destination folder. This shouldn't "take too long" because the operation is atomic -- it can't be done in an incomplete manner.

I find it really interesting that upon rebooting the app, it doesn't cause a crash anymore though..

Launching an app and moving the app while it's running is known to cause issues though. So maybe the user could be launching the old app before Sparkle moves the old app to a temporary location, so it can put the new app where the old one used to be. But I imagine your crash reports indicates what app version is in there and you can rule out if that is or is not the case.

**Comment by yingfangdu** [comment]
@zorgiepoo @pornel any suggestion that we can do to solve this issue now?

I do not get this line, we have no way to detect reinstalling is happening at boot time.  Or is there any way we can check whether the application bundle's integrity during launch? 
 "**But I imagine your crash reports indicates what app version is in there and you can rule out if that is or is not the case"**

**Comment by zorgiepoo** [comment]
If my theory is correct (it may very well not be) that the old app being launched by the user is moved by Sparkle's Autoupdate tool, then the application version in the crash reports should not be the newest version that was just deployed.

**Comment by yingfangdu** [comment]
@zorgiepoo , the crashed application is jus the one we deployed, i.e., the newest version.
